README.md
# AIM - Attack Improvement Mod 2.1 Development #
For BATTLETECH 1.1.2

- [Features Overview](#features-overview)
- [Installation](#installation)
- [Configuration](#configuration)
- [Settings](#settings)
  * [User Interface Settings](#user-interface-settings)
  * [Targeting Line Settings](#targeting-line-settings)
  * [Called Shot Settings](#called-shot-settings)
  * [Melee and DFA Settings](#melee-and-dfa-settings)
  * [Hit Modifier Settings](#hit-modifier-settings)
  * [Hit Roll Settings](#hit-roll-settings)
  * [Hit Resolution Settings](#hit-resolution-settings)
  * [Attack Log Settings](#attack-log-settings)
- [Compatibilities](#compatibilities)
- [The Story of AIM](#the-story-of-aim)
- [Learn to Mod](#learn-to-mod)


AIM is a BattleTech mod that fixes, enhances, and customise your combat experience, such as fixing vehicle called shot, improves targeting lines, tune down roll correction, show heat and stability numbers, and a detailed attack log.  The default settings aims to not deviate too far from vanilla.

This mod does *not* modify game data.  Saves made with this mod on will *not* be affected by disabling or removing this mod.

# Features Overview

**ALL features can be enabled or disabled individually.**

## Game Bugs Fixed and HUD Enhancements ##

* Vehicle Called Shot fixed.
* Damaged Structure Display fixed and enhanced.
* Grey Head Disease fixed.
* Line of Fire fixed and stylised.
* Multi-Target Back Out fixed.
* Make sure 0 HP means dead, never zombie.
* Show Heat and Stability Numbers, including cooldown overhead.
* Show Base Hit Chance in accuracy modifier popup.
* (Optional) Show Mech Tonnage.
* (Optional) Show Corrected Hit chance and Called Shot Chance.

## Mechanic Enhancements ##

* Unlock Hit Chance Stepping to make odd piloting useful.
* Called Shots cluster around called mech location.
* Precise Hit Distribution that improves SRM and MG called shot.
* More Melee Modifiers, which fixes the absent of stood up penalty.
* Allow Net Bonus Hit modifiers.
* Allow Negative Height Modifier.
* Remove Melee Positon Cap.

## Other Adjustables ##

* Tabular Attack Log that can be copied and pasted to Excel.
* Adjustable Roll Correction Strength, default halved.
* Adjustable Miss Streak Breaker.
* Adjustable Base Hit Chance.
* Adjustable Hit Chance stepping and min/max cap.
* Control Display Precision of hit chance and called shot chance.


# Installation

<span style="color:red">
**IMPORTANT Note on upgrading FROM mod version 1.0**:</span>
Rename your old `mod.json` to `settings.json` to keep old settings.

1. [Install BTML and ModTek.](https://github.com/Mpstark/ModTek/wiki/The-Drop-Dead-Simple-Guide-to-Installing-BTML-&-ModTek-&-ModTek-mods)
2. [Download this mod](https://github.com/Sheep-y/Attack-Improvement-Mod/releases), extract in the mod folder. i.e. You should see `BATTLETECH\Mods\AttackImprovementMod\mod.json`
3. Launch the game. The mod will creates a "settings.json" and a mod log in the same folder as `mod.json`.
4. Open `settings.json` to see mod settings.  If you want to change it, restart game to apply changes.


# Configuration

When the mod is first loaded by the game, it will try to read settings.json and validate its settings.

Some configurations are bundled with the mod.  You may copy or rename a setting to `settings.json` to apply it.

* `settings.default.json` - Out of box default.
* `settings.spartan.json` - Enable diminishing modifier, more melee modifiers, more info display, and disables roll correction and streak breaker.
* `settings.fix-only.json` - Only enable game fixes for an enhanced vanilla experience.  No clustered called shot or hit chance stepping unlock.
* `settings.log-only.json` - Disable all features except attack log, which is enabled in full.

Note that `settings.json` is auto-managed.  Old settings will be upgraded and removed, out of range settings will be corrected, and the formats and comments are fixed.  You can only change setting values, which may be written back to the file in a standard computerised form.

Unlike many simpler mods, this mod is designed to run as fast as it can be, so don't worry about performance.  Disabled features won't slow down the game and does not modify any code.

# Settings

These settings can be changed in `settings.json`.


## User Interface Settings


**Fix Rear Structure Display**

> Setting: `FixPaperDollRearStructure`  (true/false, default true)
>
> The rear structure of the paper dolls are showing front structure on the reverse side because of a typo in the code.  If the front has armour, it means the damaged structure will not be displayed.
>
> When set to true, this mod can fix all paper dolls, including selection panel, target panel, called shot panel, post-mission status, and in mech bay.


**Show Under-Armour Damage**

> Setting: `ShowUnderArmourDamage `  (true/false, default true)
>
> Not sure why the game didn't do this.  When set to true, armour of damaged location will have a striped pattern instead of solid.


**Show Heat and Stability Numbers**

> Setting: `ShowHeatAndStab`  (true/false, default true)
>
> When set to true, display heat and stability numbers in the selection panel (bottom left) and targeting panel (top center), and predicts post action numbers.  Predictions supplied by the game and is subject to all its quirks and bugs and mods.


**Show Unit Tonnage**

> Setting: `ShowUnitTonnage`  (true/false, default false)
>
> When set to true, show mech and vehicle tonnage in selection and target panel.
>
> Duplicates with Extended Information, but AIM override it and use a shorter form for mechs to fit in heat and stability numbers.  Default false because the short form can overwhelm inexperienced players.


**Show Ammo In Tooltip**

> Setting: `ShowAmmoInTooltip`  (true/false, default true)<br>
> Setting: `ShowEnemyAmmoInTooltip`  (true/false, default false)<br>
>
> When set to true, show ammo count in the list of components when you mouseover a location on the paper doll.
>
> The main purpose is to allow you to see the state of each ammo bin and tell whether they are at risk of exploding.


**Fix Multi-Target Back Out**

> Setting: `FixMultiTargetBackout`  (true/false, default true)
>
> The game's Muti-Target back out (escape/right click) is bugged. Backing out from first target will cancel the action, and second back out (regardless of target) will always cancel the whole thing.
>
> When set to true, this mod will make Multi-Target back out deselect targets one by one as expected.



## Targeting Line Settings

**Change Widths of Targeting Lines**

> Setting: `LOSWidth`  (0 to 10, default 2, game default 1)
>
> Set width of all targeting lines (direct, indirect, blocked, can't attack etc.).  Game default is 1  Mod default is 2.
<br>

> Setting: `LOSWidthBlocked`  (0 to 10, default 1.5, game default 0.75)
>
> Set width of obstructed part of an obstructed targeting lines, which is normally thinner than other lines by default.  Game default is 0.75.  Mod default is 1.5.
>
> When the mod "Firing Line Improvement" is detected, this setting will be disabled to avoid conflicts.


**Styles and Colours Targeting Lines**

> Setting: `LOSIndirectDotted`  (default true, game default false)<br>
> Setting: `LOSNoAttackDotted`  (default true)<br>
> Setting: `LOSMeleeDotted`  (default false)<br>
> Setting: `LOSClearDotted`  (default false)<br>
> Setting: `LOSBlockedPreDotted`   (default false)<br>
> Setting: `LOSBlockedPostDotted`  (default false)<br>
> Setting: `LOSMeleeColor`  (default "")<br>
> Setting: `LOSClearColor`  (default "")<br>
> Setting: `LOSBlockedPreColor`   (default "#D0F")<br>
> Setting: `LOSBlockedPostColor`  (default "#C8E")<br>
> Setting: `LOSIndirectColor`  (default "")<br>
> Setting: `LOSNoAttackColor`  (default "")
>
> Set the colour and style of various targeting lines.
> Obstructed lines has two parts. The part before obstruction is Pre, and the part after is Post.
>
> Colours are either empty or in HTML hash syntax.  For example `"#F00"` = red, `"#0F0"` = green, `"#00F"` = blue, `"#FFF"` = white, `"#888"` = grey, `"#000"` = black.
> Four parts means RGBA, while three parts mean full opacity RGB.  Supports full and short form. e.g. #28B = #2288BB = #2288BBFF.
>
> When the mod "Firing Line Improvement" is detected, this setting will be disabled to avoid conflicts.


**Change Widths of Obstruction Marker**

> Setting: `LOSMarkerBlockedMultiplier`  (0 to 10, default 1.5)
>
> Scale the obstruction marker of targeting lines, the "light dot" that split the obstructed line into two. 2 means double width and height, 0.5 means half-half.  Set to 1 to leave at game default.  Set to 0 will not remove them from game but will effectively hide them.
>
> When the mod "Firing Line Improvement" is detected, this setting will be disabled to avoid conflicts.


**Refine or Rough Fire Arc and Jump Arc**

> Setting: `ArcLinePoints`  (2 to 1000, default 48, game default 18)
>
> On a high resolution monitor it is easy to see the hard corners of the arc of indirect targeting lines.  Lines are quick to draw, so this mod will happily improves their qualities for you.
> Set to 2 to make them flat like other lines.  Set to 18 to leave at game default.
>
> When the mod "Firing Line Improvement" is detected, this setting will be disabled to avoid conflicts.


**Fix LoS Inconsistency between Walk and Jump**

> Setting: `FixLosPreviewHeight`  (true/false, default true)
>
> Walk and Jump will sometimes predicts different Line of Sight, because the walking preview height is slightly different from jumping preview height.  Set to true to make them the same.



## Called Shot Settings

**Enable Vehicle Called Shot**

> Setting: `FixVehicleCalledShot`  (true/false, default true)
>
> Called shot have absolutely no effect on vehicles, as of game version 1.1.2.  Set to true to make it work again.
>
> In the code, Mech Locations and Vehicle Locations are two separate data type, meaning there are two set of states and methods to handle hit location.  And vehicle's called shot handling is incomplete.  You can pick a called shot location, but it never reach the hit location routine.  This can be verified by disabling this setting and enabling attack log.



**Fix Grey Head Disease**

> Setting: `FixGreyHeadDisease`  (true/false, default true)
>
> Set to true to confine the grey head disease to the boss and not spread.
>
> When anyone, friends or foes, attacks a headshot immune character, all attacks from the same direction will never hit the head ever again.  Every one's head will be grey.  I call it the grey head disease.  It lasts until you load the game.
<br>

> Setting: `FixBossHeadCalledShotDisplay`  (true/false, default true)
>
> Set to true to make sure boss head is always unselectable.
>
> Did you know that grey head in called shot popup is actually a bug?
(I hope it is a bug.)
> If you try to call shot the boss before any headshot-immune unit is attacked, such as right after a load, the head is actually selectable!
> This is most apparent with FixGreyHeadDisease on, since the head will be always available.
>
> If `FixGreyHeadDisease` is true but `FixBossHeadCalledShotDisplay` is false, the boss's head will be selectable for called shot, but it is false hope.  Your shots will never hit the head.


**Enable Clustering Called Shot**

> Setting: `CalledShotUseClustering` (true/false, default true)
>
> When set to true, called shot has a higher chance to hit adjacent locations.
>
> For example, head called shot would bias the head, but also the three torsos to a lesser degree.
>
> This is the default behaviour on and before game version 1.0.4, which was bugged and caused very low called head shot chances since head is excluded from clustering.  The bug is one of the driving forces of this mod's initial creation.  This mod can recreate the clustering effect without the bug.
>
> Note that this does not apply to Vehicle called shot; the default vehicle clustering is only half done, and even when fixed there are too few locations to have meaningful cluster.



**Adjust Called Shot Weight**

> Setting: `MechCalledShotMultiplier`  (0 to 1024.0, default 0.33)
>
> When clustering called shot is enabled, chance to hit called locations will be amplified by clustering weight.  This setting let you tune called shot's weight multipliers.
> Default is 0.33 to counter the effect of CalledShotClusterStrength.  Set to 1.0 would leave original multiplier unchanged, while 0.0 will removing it leaving only clustering effect (if enabled).
<br>

> Setting: `VehicleCalledShotMultiplier`  (0 to 1024.0, default 0.75)
>
> Called shot didn't work on vehicles without modding.  Once that is fixed, unmodified called shot is pretty powerful on vehicles because of its low number of locations.  This setting tries to balance that.


**Show Modded Called Shot Chance**

> Setting: `ShowRealMechCalledShotChance`  (true/false, default true)<br>
> Setting: `ShowRealVehicleCalledShotChance`  (true/false, default true)
>
> When enabled, the popups will reflect modded hit distribution.  Also, if vehicle called shot fix is disabled, vehicle called shot chance will be updated to reflect disabled called shot.


(Advanced) **Format Called Shot Hit Distribution Chances**

> Setting: `CalledChanceFormat`  (string, default "")
>
> Use Microsoft C# [String.Format](https://docs.microsoft.com/en-us/previous-versions/visualstudio/visual-studio-2008/0c899ak8(v=vs.90) syntax to format called shot location chances.
>
> Set to "{0:0.0}%" to always show one decimal, or "{0:0.00}%" for two decimals.
> Leave empty to round them to nearest integer.
>
> Replace the old "ShowDecimalCalledChance" setting in mod version 1.0.


## Melee and DFA Settings


**Allow free Melee and DFA positioning**

> Setting: `IncreaseMeleePositionChoice`  (true/false, default true)<br>
> Setting: `IncreaseDFAPositionChoice`  (true/false, default true)
>
> When set to true, melee and DFA can use all available positions, instead of nearest three.  Compatible with MeleeMover.
>
>
> Setting: `UnlockMeleePositioning`  (true/false, default true)
>
> When enabled, free player units from standing still when target is already in melee range.  Auto-disable to avoid conflict when MeleeMover is used.


**Change Melee Modifiers**

> Setting: `MeleeAccuracyFactors`  (comma separated value)
>
> A list of hit modifiers of melee and DFA attacks. Leave empty to keep it unchanged.  Order will be followed, and letter case does not matter.
>
> Since this feature will override both mouseover display and actual modifier calculation, this will fix the bug that SelfStoodUp is displayed in mouseover but not counted in modifier.
>
> Default is "DFA, Height, Inspired, SelfChassis, SelfHeat, SelfStoodUp, SelfWalked, Sprint, TargetEffect, TargetEvasion, TargetProne, TargetShutdown, TargetSize, TargetTerrainMelee, WeaponAccuracy".
>
> Other options are ArmMounted, Obstruction, Refire, SelfTerrain, SensorImpaired, and TargetTerrain.
>
> **ArmMounted** - Apply arm mount bonus if the punching arm is intact and the attack is not DFA and not against prone mech or vehicle. <br>
> **DFA** - Apply DFA penalty if attack is DFA. <br>
> **Height** - Apply one level of height modifier if height different is at least half of melee reach.  DFA height difference is calculated like ranged weapon - from pre-flight attacker position to target position. <br>
> **Inspired** - Apply inspired bonus. <br>
> **Obstruction** - Apply obstructed penalty. <br>
> **Refire** - Apply refire penalty.  Which should be 0 by default but can be changed in weapon data files. <br>
> **SelfChassis** - Apply chassis modifier. <br>
> **SelfHeat** - Apply overheat penalty. <br>
> **SelfStoodUp** - Apply stood up penalty. <br>
> **SelfTerrain** - Apply self terrain penalty as if this is a ranged attack. <br>
> **SelfWalked** - Apply self walked penalty, default 0 but can be changed in game configuration file. <br>
> **SensorImpaired** - Apply sensor impaired penalty. <br>
> **Sprint** - Apply sprint penalty. <br>
> **TargetEffect** - Apply target effect penalty such as gyro. <br>
> **TargetEvasion** - Apply target evasion penalty.  Melee attacks ignore up to 4 evasion by default. <br>
> **TargetProne** - Apply target prone penalty. <br>
> **TargetShutdown** - Apply target shutdown bonus. <br>
> **TargetSize** - Apply target size penalty. <br>
> **TargetTerrain** - Apply target terrain's modifier as if this is a ranged attack. <br>
> **TargetTerrainMelee** - Apply target terrain's melee modifier. <br>
> **WeaponAccuracy** - Apply weapon accuracy modifier. <br>



## Hit Modifier Settings

**Allow Net Bonus Modifier**

> Setting: `AllowNetBonusModifier`  (true/false, default true)
>
> When set to true, total modifier of an attack can be a net bonus that increases the hit chance beyond the attacker's base hit chance (but still subjects to 95% cap unless lifted by the `MaxFinalHitChance` settings).
> Game default is false.
>
> When the net modifier is a bonus, it will use the same handling as penalty but reversed: first 10 modifiers are ±5% each, and subsequence modifiers are ±2.5% each.


**Allow Height Diff Penalty**

> Setting: `AllowLowElevationPenalty`  (true/false, default true)
>
> When set to true, attacking from low ground to high ground will incur an accuracy penalty that is the exact reverse of attacking from high ground to low.
> Game default is false.


**Modify Base Hit Chance**

> Setting: `BaseHitChanceModifier` (-10.0 to 10.0, default 0)<br>
> Setting: `MeleeHitChanceModifier` (-10.0 to 10.0, default 0)
>
> Increase or decrease base hit chance of ranged/melee attacks.
> e.g. -0.05 to lower base accuracy by 5%, 0.1 to increase it 10%.


**Unlock Modifier Stepping and Range**

> Setting: `HitChanceStep`  (0.0 to 1.0, default 0)
>
> The game will round down final hit chance to lower 5% by default.
> This affects some calculations, such as rendering odd gunnery stats and piloting stats less effective then they should be.
> Set this to 0 to remove all stepping.  Set it to 0.005 will step the accuracy by 0.5%, and so on.
>
>
> Setting: `MaxFinalHitChance`  (0.1 to 1.0, default 0.95)<br>
> Setting: `MinFinalHitChance`  (0.0 to 1.0, default 0.0)
>
> Use this to set max and min hit chance after all modifiers but before roll correction.
> Note that 100% hit chance may still miss if roll correction is enabled.

**Diminishing Hit Chance Modifier**

> Setting: `DiminishingHitChanceModifier`  (true/false, default false)
>
> Set this to true to enable diminishing return of modifiers, instead of simple add and subtract.
> As a result, small penalties have a bigger effect, but very large penalties may be more bearable.
<br>

> Setting: `DiminishingBonusPowerBase`  (default 0.8)<br>
> Setting: `DiminishingBonusPowerDivisor`  (default 6)<br>
> Setting: `DiminishingPenaltyPowerBase`  (default 0.8)<br>
> Setting: `DiminishingPenaltyPowerDivisor`  (default 3.3)<br>
>
> Bonus formula is "2-Base^(Modifier/Divisor)". <br>
> Example: +3 Bonus = 0.8^(3/6) = -1.1055728 = 110%. <br>
> Thus +3 Bonus @ 80% To Hit = 80% x 110% = 88% final to hit.
>
> Penalty formula is "Base^(Modifier/Divisor)". <br>
> Penalty Example: +6 Penalty = 0.8^(6/3.3) = 0.6665 = 66.7%. <br>
> Thus +6 Penalty @ 80% To Hit = 80% x 66.7% = 53% final to hit.
<br>

> Setting: `DiminishingBonusMax`  (default 16)
> Setting: `DiminishingPenaltyMax`  (default 32)
>
> The modifiers are pre-calculated to run faster.  These settings determine how many results are cached.  Modifiers beyond the max will be regarded as same as max.


## Hit Roll Settings

**Adjust Roll Correction**

> Setting: `RollCorrectionStrength`  (0.0 to 2.0, default 0.5)
>
> It is no secret that the game fudge all hit rolls, called a "correction".  As a result, real hit chances are shifted away from 50%, for example 75% becomes 84% while 25% becomes 16%.  This can create a rift between what you see and what you get, especially on low chance shots.
>
> This mod does not aim to completely disable roll adjustment, and thus default to half its strength.  You can set the strength to 0 to disable it, 1 to use original formula, 2 to amplify it, or any value between 0 and 2.
>
> If the "True RNG Hit Rolls" mod is detected, this setting will be switched to 0 for consistency.


(Advanced) **Adjust Miss Streak Threshold**

> Setting: `MissStreakBreakerThreshold`  (0.0 to 1.0, default 0.5)
>
> In addition to roll adjustment, the game also has a "miss streak breaker".  Whenever you miss an attack of which uncorrected hit chance > 50%, the streak breaker will adjust your hit chance up on top of roll correction.  The bonus accumulates until you land a hit (regards of hit chance), at which point it resets to 0.
>
> This setting let you adjust the threshold.  0.75 means it applies to attack of which hit chance > 75% (excluding 75%).  Set to 0 enable it for all attacks, or set to 1 to disable it.  Default is 0.5 which is the game's default.
>
> If the "True RNG Hit Rolls" mod is detected, this setting will be switched to 1 for consistency.


(Advanced) **Adjust Miss Streak Bonus**

> Setting: `MissStreakBreakerDivider`  (-100.0 to 100.0, default 5.0)
>
> For every miss that crosses the streak breaker threshold, the threshold is deduced from hit chance, then divided by 5.  The result is then added as streak breaker bonus.
>
> Set this setting to a positive number to override the divider.  For example at threshold 0.5 and divider 3, a 95% miss result in (95%-0.5)/3 = 15% bonus to subsequence shots until hit.  Default is 5 which is the game's default.
>

changelog.txt
CHANGE LOG

2018-07-26 AIM Version 2.0 for BATTLETECH 1.1.2

* Mod: Settings migrated to settings.json and is automatically updated by the mod.
* New: Unlock hit accuracy's 5% step and make odd gunnery useful.
* New: Show current and projected heat and stability number in selection panel and target panel.
* New: Fix and allow to-hit modifier to be a net bonus, instead of capping at no penalty.
* New: Allow height modifier to be net penalty, instead of capping at no bonus.
* New: Display base hit chance of ranged and melee attack and the relevant stat.
* New: Fix Multi-Target back out.
* New: Fix rear armour structure display, and show structural damage through armour.
* New: Fix Grey Head Disease, spread by headshot immune unit to all mechs until load game or restart.
* New: Fix the bug that parts and units has a very low chance to be zombified at 0 HP.
* New: Make sure walk and jump result in same LOS quality (clear/blocked/indirect/none).
* New: Change Line of Fire colours, style (solid/dotted), and width.
* New: Refine (or rough up / flatten) indirect firing arc and jump arc.
* New: Unlock melee positioning: allow all valid spots to be moved to and mechs are not locked in place by melee.
* New: Customise list and order of melee attack modifiers, with new factors such as height and inspired.<br>When used, fix the bug that prone stand up penalty is not actually applied to melee.
* New: A rewritten attack log that is faster and supports various level of details including direction, damage, and crit.
* New: (Optional) Change ranged/melee base hit chances and min/max of modified hit chance.
* New: (Optional) Diminishing to-hit modifier effect that makes light penalties more punishing and big penalties less hopeless.
* New: (Optional) Show mech and vehicle tonnage in selection panel and target panel, before unit class.
* Mod: New mod architecture that is faster, safer, and enables mod conflict detection and resolution.
* Mod: Detect and resolve conflicts with Firing Line Improvement, Extended Information, Melee Mover, True RNG, and Real Hit Chance.
* Mod: Log folder can now be set through the LogDir setting.
* Mod: More setting validations.
* Fix: Compatibility with early mod loading, before most game states are initiated.
* Fix: Re-enable hit distribution override by default.
* Fix: Display of corrected hit chance no longer caps at 95%.
* Fix: Call chance displays no longer inter-dependent; enabling any relevant settings will triggers display patch.
* Fix: Hit chance format is now free, and defaults to optionally displayed to one decimal point if stepping modifier is used and 5% stepping is removed.
* Fix: Attack Log now logs attacks on turrets and buildings, and fill in all blanks. Total weight and call goal removed.
* Fix: Mod log now ignores all duplicate exceptions, if logged through Error( object ), to not slow down the game with disk writes.
* Fix: Optimise miss streak breaker bypass when it is disabled. (Roll correction already has fast bypass.)

2018-06-28 AIM Version 1.0.1 for BATTLETECH 1.1.1

* Fix: Fix vehicle called shot no longer float mech location name.
* Fix: Auto-detect Log directory.
* Fix: Cap accuracy between 0 and 100.
* Fix: Override called shot percent cap when ShowRealWeaponHitChance is true but ShowDecimalHitChance is false.
* Fix: Auto-bypass Hit Location bug fix on 1.1.1 and up.

2018-06-21 AIM Version 1.0.0 for BATTLETECH 1.0 and 1.1

Initial version with 
* Vehicle Called Shot fixed
* Hit Location fixed
* Clustering Called Shot
* Roll Correction Adjustment
* Miss Streak Adjustment
* Attack Roll Log.


AttackImprovementMod/RollCorrection.cs
using BattleTech;
using BattleTech.UI;
using System;

namespace Sheepy.AttackImprovementMod {
   using static Mod;
   using System.Reflection;
   using System.IO;
   using UnityEngine;

   public class RollCorrection {

      internal const string ROLL_LOG = "Log_AttackRoll.txt";
      private static bool DisableRollCorrection = false;
      private static readonly float[] correctionCache = new float[21];

      internal static void InitPatch () {
         if ( ! Settings.PersistentLog ) DeleteLog( ROLL_LOG );
         DisableRollCorrection = Settings.RollCorrectionStrength == 0.0f;
         Type AttackType = typeof( AttackDirector.AttackSequence );

         FieldInfo rollCorrection = typeof( AttackDirector.AttackSequence ).GetField( "UseWeightedHitNumbers", BindingFlags.Static | BindingFlags.NonPublic );
         bool rollCorrected = true;
         if ( rollCorrection == null )
            Log( "Warning: Cannot find AttackDirector.AttackSequence.UseWeightedHitNumbers." );
         else 
            rollCorrected = (bool) rollCorrection.GetValue( null );

         if ( Settings.LogHitRolls ) {
            Patch( AttackType, "GetIndividualHits", BindingFlags.NonPublic | BindingFlags.Instance, "RecordAttacker", null );
            Patch( AttackType, "GetClusteredHits" , BindingFlags.NonPublic | BindingFlags.Instance, "RecordAttacker", null );
            Patch( AttackType, "GetCorrectedRoll" , BindingFlags.NonPublic | BindingFlags.Instance, new Type[]{ typeof( float ), typeof( Team ) }, "RecordAttackRoll", "LogMissedAttack" );
            if ( ! File.Exists( ROLL_LOG ) )
               RollLog( String.Join( "\t", new string[]{ "Attacker", "Weapon", "Hit Roll", "Corrected", "Streak", "Final", "To Hit", "Location Roll", "Head/Turret", "CT/Front", "LT/Left", "RT/Right", "LA/Rear", "RA", "LL", "RL", "Called Part", "Called Bonus", "Total Weight", "Goal", "Hit Location" } ) );
         }

         if ( DisableRollCorrection ) {
            if ( ! rollCorrected ) {
               Log( "Roll correction is already disabled." );
            } else {
               rollCorrection.SetValue( null, false );
            }
            rollCorrected = false;

         } else {
            if ( Settings.RollCorrectionStrength < 0 ) {
               Log( "Error: RollCorrectionStrength must not be negative." );
               Settings.RollCorrectionStrength = 1.0f;
            } else if ( Settings.RollCorrectionStrength > 1.9999f ) {
               if ( Settings.RollCorrectionStrength > 2 )
                  Log( "Warning: RollCorrectionStrength must be less than 2." );
               Settings.RollCorrectionStrength = 1.9999f; // Max! 1.99999 results in NaN in reverse correction
            }

            if ( Settings.RollCorrectionStrength != 1.0f )
               Patch( AttackType, "GetCorrectedRoll", BindingFlags.NonPublic | BindingFlags.Instance, new Type[]{ typeof( float ), typeof( Team ) }, "OverrideRollCorrection", null );
            if ( rollCorrected && Settings.ShowRealWeaponHitChance ) {
               for ( int i = 0 ; i < 21 ; i++ )
                  correctionCache[i] = ReverseRollCorrection( 0.05f * i, Settings.RollCorrectionStrength );
               Patch( typeof( CombatHUDWeaponSlot ), "SetHitChance", typeof( float ), "ShowRealHitChance", null );
            }
         }
         if ( Settings.MissStreakBreakerThreshold != 0.5f || Settings.MissStreakBreakerDivider != 5f ) {
            StreakBreakingValueProp = typeof( Team ).GetField( "streakBreakingValue", BindingFlags.NonPublic | BindingFlags.Instance );
            if ( StreakBreakingValueProp != null )
               Patch( typeof( Team ), "ProcessRandomRoll", new Type[]{ typeof( float ), typeof( bool ) }, "OverrideMissStreakBreaker", null );
            else
               Log( "Error: Can't find Team.streakBreakingValue. Miss Streak Breaker cannot be patched." );
         }

         if ( Settings.ShowDecimalHitChance ) {
            Patch( typeof( CombatHUDWeaponSlot ), "SetHitChance", typeof( float ), "OverrideWeaponHitChanceFormat", null );
            if ( ! Settings.ShowRealWeaponHitChance )
               Log( "Warning: ShowDecimalHitChance without ShowRealWeaponHitChance" );
         } else if ( Settings.ShowRealWeaponHitChance ) {
            Patch( typeof( CombatHUDWeaponSlot ), "SetHitChance", typeof( float ), "OverrideWeaponHitChanceRange", null );
         }
      }

      // ============ UTILS ============

      internal static void RollLog ( String message ) {
         WriteLog( ROLL_LOG, message + "\r\n" );
      }

      public static float CorrectRoll( float roll, float strength ) {
         strength /= 2;
         return (float)( (Math.Pow(1.6*roll-0.8,3)+0.5)*strength + roll*(1-strength) );
      }

      // A reverse algorithm of AttackDirector.GetCorrectedRoll
      internal static float ReverseRollCorrection ( float target, float strength ) {
         if ( strength == 0.0f ) return target;
         // Solving r for target = ((1.6r-0.8)^3+0.5)*(s/2)+r*(1-s/2)
         double t = target, t2 = t*t, s = strength, s2 = s*s, s3 = s2*s,
                a = 125 * Math.Sqrt( ( 13824*t2*s - 13824*t*s - 125*s3 + 750*s2 + 1956*s + 1000 ) / s ),
                b = a / ( 4096*Math.Pow( 6, 3d/2d )*s ) + ( 250*t - 125 ) / ( 1024 * s ),
                c = Math.Pow( b, 1d/3d );
         return c == 0 ? target : (float)( c + (125*s-250)/(1536*s*c) + 0.5 );
      }

      // ============ Fixes ============

      public static bool OverrideRollCorrection ( ref float __result, float roll, Team team ) {
         try {
            roll = CorrectRoll( roll, Settings.RollCorrectionStrength );
            if ( team != null )
               roll -= team.StreakBreakingValue;
            __result = roll;
            return false;
         } catch ( Exception ex ) {
            return Log( ex );
         }
      }

      private static FieldInfo StreakBreakingValueProp = null;
      public static bool OverrideMissStreakBreaker ( Team __instance, float targetValue, bool succeeded ) {
         try {
            if ( succeeded ) {
               StreakBreakingValueProp.SetValue( __instance, 0f );

            } else if ( targetValue > Settings.MissStreakBreakerThreshold ) {
               float mod;
               if ( Settings.MissStreakBreakerDivider > 0 )
                  mod = ( targetValue - Settings.MissStreakBreakerThreshold ) / Settings.MissStreakBreakerDivider;
               else
                  mod = - Settings.MissStreakBreakerDivider;
               StreakBreakingValueProp.SetValue( __instance, __instance.StreakBreakingValue + mod );
            }
            return false;
         } catch ( Exception ex ) {
            return Log( ex );
         }
      }

      public static void ShowRealHitChance ( ref float chance ) {
         chance = Mathf.Clamp( chance, 0f, 1f );
         int i = (int)( ( chance + 0.00001f ) / 0.05f );
         if ( Math.Abs( i * 0.05f - chance ) < 0.00001f )
            chance = correctionCache[ i ];
         else {
            Log( "Uncached hit chance reversal from " + chance + ", diff: " + ( i * 0.05f - chance ) );
            chance = ReverseRollCorrection( chance, Settings.RollCorrectionStrength );
         }
      }

      private static MethodInfo HitChance = typeof( CombatHUDWeaponSlot ).GetMethod( "set_HitChance", BindingFlags.Instance | BindingFlags.NonPublic );
      private static MethodInfo Refresh = typeof( CombatHUDWeaponSlot ).GetMethod( "RefreshNonHighlighted", BindingFlags.Instance | BindingFlags.NonPublic );
      private static readonly object[] empty = new object[]{};

      // Override the original code to show accuracy in decimal points
      public static bool OverrideWeaponHitChanceRange ( CombatHUDWeaponSlot __instance, float chance ) {
         try {
            HitChance.Invoke( __instance, new object[]{ chance } );
            __instance.HitChanceText.text = string.Format( "{0:0}%", Mathf.Clamp( chance * 100f, 0f, 100f ) );
            Refresh.Invoke( __instance, empty );
            return false;
         } catch ( Exception ex ) {
            return Log( ex );
         }
      }

      // Override the original code to show accuracy in decimal points
      public static bool OverrideWeaponHitChanceFormat ( CombatHUDWeaponSlot __instance, float chance ) {
         try {
            HitChance.Invoke( __instance, new object[]{ chance } );
            __instance.HitChanceText.text = string.Format( "{0:0.0}%", Mathf.Clamp( chance * 100f, 0f, 100f ) );
            Refresh.Invoke( __instance, empty );
            return false;
         } catch ( Exception ex ) {
            return Log( ex );
         }
      }

      // ============ Log ============

      // Get attacker, weapon and hitchance before logging
      internal static string thisAttacker = "(unknown)";
      internal static string thisWeapon = "(unknown)";
      internal static float thisHitChance;
      public static void RecordAttacker ( AttackDirector.AttackSequence __instance, Weapon weapon, float toHitChance ) {
         thisAttacker = __instance.attacker.GetPilot()?.Callsign ?? __instance.attacker.Nickname;
         thisWeapon = weapon.defId.StartsWith( "Weapon_" ) ? weapon.defId.Substring( 7 ) : weapon.defId;
         thisHitChance = toHitChance;
      }

      internal static float thisRoll;
      internal static float thisStreak;
      public static void RecordAttackRoll ( float roll, Team team ) {
         thisRoll = roll;
         thisStreak = team?.StreakBreakingValue ?? 0;
      }

      internal static float thisCorrectedRoll;
      public static void LogMissedAttack ( float __result, float roll, Team team ) {
         thisCorrectedRoll = __result;
         if ( __result > thisHitChance ) // Miss, log now because hit location won't be rolled
            RollLog( GetHitLog() +
               "\t--" + // Roll; Empty cells are added so that copy and paste will override any old data in Excel, instead of leaving them in place and make it confusing
               "\t--\t--\t--\t--" +  // Head & Torsos
               "\t--\t--\t--\t--" +  // Limbs
               "\t--\t--\t--\t--" ); // called shot and result
      }

      internal static string GetHitLog () {
         return thisAttacker + "\t" + thisWeapon + "\t" + thisRoll + "\t" + ( thisCorrectedRoll + thisStreak ) + "\t" + thisStreak + "\t" + thisCorrectedRoll + "\t" + thisHitChance;
      }
   }
}